
\chapter{Conclusions}
\label{Chapter5}

\textit{Rewind: short recap}

It is of particular biological interest that this method provides extremely flexible abundance estimation capabilities: by tuning the regularization parameters, it is possible to match the complexity of different metagenomic experiments in real time, without the need to re-run costly simulation or learning steps. This is extremely useful in situations where the compositions of the samples are completely unknown, since it enables their analysis using an extensive database of reference genomes and provides accurate estimates of the abundances of each taxonomic clade within minutes, as well as samples of well-known expected composition and complexity, where the task is to accurately estimate genome-level abundances to compare them to other samples.

Nevertheless, there are some challenges still to be addressed: first, genomes present in very low abundances are very difficult to detect, since their coefficients are likely to be annihilated during the optimization of the regularized model. On the other hand, reducing the regularization parameters may have an even worse effect on the results, introducing hundreds of low-abundant false positives indistinguishable from those actually present in low abundances. This issue is intrinsic to the nature of microbial genomes, subject to phenomena such as horizontal gene transfer thanks to which prokaryotic organisms of different species can exchange portions of their genetic material, decreasing the probability that reads can be assigned to the exact reference genome they were generated from, in the current state of read sequencing technologies, producing short reads which may map equally well to tens or hundreds of genomes, and in the heuristic assignment algorithms which are unavoidable if we want to make the analysis of metagenomes computationally feasible. The scientific literature confirms that it is unlikely to obtain accurate genome-level abundances for real-sized metagenomic samples and reference datasets [ref lindgreen, growrs, ditasic], without restricting the search to a tractable number of ``genomes of interest''.

\textit{High-impact closing-up paragraph}
